### PR TITLE
Lang: feature flag init if needed

### LINKED ISSUE #1191: lang: put `init_if_needed` behind feature flag

### PR DESCRIPTION
**Pull request by paul-schaaf**
closes https://github.com/project-serum/anchor/issues/1191

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -11,10 +11,14 @@ incremented for features.
...
 ## [Unreleased]
 
-#### Fixes
+### Fixes
 
 *lang: Improved error msgs when required programs are missing when using the `init` constraint([#1257](https://github.com/project-serum/anchor/pull/1257))
 
+### Features
+
+*lang: Put `init_if_needed` behind a feature flag to decrease wrong usage ([#1258](https://github.com/project-serum/anchor/pull/1258))
```
Comment (armaniferrante) [inline CHANGELOG.md:20]: Technically this is breaking. :shrug:

**Comment by armaniferrante** [comment]
Blocked by https://github.com/project-serum/anchor/pull/1199.

**Comment by paul-schaaf** [comment]
@armaniferrante any idea what might be causing the test to fail? Im not able to reproduce it locally.

File: lang/syn/src/parser/accounts/constraints.rs [hunk 2]
```diff
@@ -344,6 +344,20 @@ impl<'ty> ConstraintGroupBuilder<'ty> {
     pub fn build(mut self) -> ParseResult<ConstraintGroup> {
         // Init.
         if let Some(i) = &self.init {
+            #[cfg(not(feature = "init-if-needed"))]
```
Comment (armaniferrante) [inline lang/syn/src/parser/accounts/constraints.rs:347]:
```suggestion
            if cfg!(not(feature = "init-if-needed"))
```

Comment (armaniferrante) [reply lang/syn/src/parser/accounts/constraints.rs:347]: testing this

**Comment by ceciEstErmat** [comment]
I use it to initialize escrows of mints "on demand".  Let's say a user wants to escrow mint X if it is the first one to do so then it will create the program escrow token account for this mint else it would just use the token account.

I found that feature pretty handy ! Even tho is forces us to always pass the system programs and token program and mint account 🤔

**Comment by paul-schaaf** [comment]
you can still do that. The feature isnt gone. You just have to explicitly enable it

**Comment by arcticmatt** [comment]
maybe add some documentation about how to enable it on this PR?

**Comment by ceciEstErmat** [comment]
@arcticmatt do something like this in your cargo 
```
[dependencies]
anchor-lang = {version = "0.23.0", features = ["init-if-needed"]}
```

**Comment by arcticmatt** [comment]
@ceciEstErmat thanks! is that any different than doing this?

```
[features]
init-if-needed = []
```

**Comment by nukec** [comment]
this is my struct:
```
#[derive(Accounts)]
#[instruction(amount: u64, bump: u8)]
pub struct UserDepositSpl<'info> {
    /// CHECK: This is not dangerous because we don't read or write from this account
    #[account(mut, signer)]
    user: AccountInfo<'info>,
    /// CHECK: This is not dangerous because we don't read or write from this account
    #[account(mut, seeds=[&user.key.as_ref()],bump=bump)]
    user_program_account: AccountInfo<'info>,
    pub system_program: Program<'info, System>,
    pub token_program: Program<'info, Token>,
    #[account(mut)]
    pub sender_token_account: Account<'info, TokenAccount>,
    #[account(mut)]
    pub pda: AccountInfo<'info>,
    #[account(
        init_if_needed,
        payer = user,
        seeds=[b"wallet".as_ref(), user.key.as_ref(), mint.key().as_ref()],
        bump = user_account_bump,
        token::mint = mint,
        token::authority = pda,
    )]
    wallet: Account<'info, TokenAccount>,
}
```

I am getting constant errors if i try to use any statement on account, whether it is `init_if_needed`, `payer`...:
` no method named `exit` found for struct `UserDepositSpl` in the current scope`

**Pull request merged**